### DISCUSSION TITLE
Automatic Argparse UI

### DISCUSSION THREAD

**Discussion by chapmanjacobd** [Textualize/textual#228] [Ideas]
It might be cool to create a class which would translate the parser or an object of parsed args into an automatic UI using the automatic Grid layout or something with a "Run" button at the lower right.

The scenario is that instead of passing a bunch of CLI arguments and flags one would be able to run the program without flags or arguments and they would be greeted with an interactive UI where they can see all the options and point, click, and type to set all the required and optional fields of a CLI tool.

maybe that user interface would look something like this:

```py
import argparse
from textual.auto import ArgParseUI

parser = argparse.ArgumentParser(add_help=False)
parser.add_argument('--foo', action='store_true', help='foo help')
parser.add_argument('bar', help='bar help')
parser.add_argument('y', nargs='?')
parser.add_argument('z', nargs='?', default='ZZ')
group = parser.add_argument_group('foo')
group.add_argument("--foo")
group.add_argument("--bar", type=int, default=42)
parser.add_argument('-req_opt', type=int, required=True)
parser.add_argument('need_one', type=str, nargs='+')

ArgParseUI(parser).run(title="argparse demo", log="arpgarse_demo.log")

# OR

try:
    args = parser.parse_args()
    run_app(args)
except:
    ArgParseUI(parser).run(title="argparse demo", log="arpgarse_demo.log")
```

https://github.com/python/cpython/blob/3.10/Lib/test/test_argparse.py

It would probably be easier to pass in parsed args instead of the parser but a more complete solution would probably need to [read the parser](https://github.com/python/cpython/blob/633db1c4eb863a1340e45c353e36f2f8dcf5945c/Lib/argparse.py#L1824) because some arguments might be `required=True` which would cause the program to halt before getting to the `textual` code. 

Maybe there could be a compromise solution using `default=argparse.SUPPRESS` or requiring people to use `nargs="?"` in all their positional arguments.

**Comment by o-murphy** [comment]
You can wraparound argparse.parse_args(sys.argv[1:]) when y you need to run you GUI before argparse handle the error of required args
